tests/scalers/temporal/temporal_test.go

```diff
@@ -0,0 +1,265 @@
+  triggers:
+  - type: temporal
+    metadata:
+      namespace: default
+      queueName: hello-task-queue
+      targetQueueSize: "2"
+      activationTargetQueueSize: "3"
+      endpoint: {{.TemporalDeploymentName}}.{{.TestNamespace}}.svc.cluster.local:7233
+`
+
+	deploymentTemplate = `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{.DeploymentName}}
+  namespace: {{.TestNamespace}}
+  labels:
+    app: {{.DeploymentName}}
+spec:
+  replicas: 0
+  selector:
+    matchLabels:
+      app: {{.DeploymentName}}
+  template:
+    metadata:
+      labels:
+        app: {{.DeploymentName}}
+    spec:
+      containers:
+      - name: nginx
+        image: nginxinc/nginx-unprivileged
+        ports:
+        - containerPort: 80
+`
+
+	jobWorkerTemplate = `
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: worker
+  namespace: {{.TestNamespace}}
+spec:
+  template:
+    spec:
+      containers:
+      - name: worker
+        image: "prajithp/temporal-sample:1.0.0"
+        imagePullPolicy: Always
+        env:
+        - name: TEMPORAL_ADDR
+          value: {{.TemporalDeploymentName}}.{{.TestNamespace}}.svc.cluster.local:7233
+        - name: TEMPORAL_NAMESPACE
+          value: default
+        - name: MODE
+          value: WORKER
+      restartPolicy: OnFailure
+  backoffLimit: 4
+`
+
+	jobWorkeFlowTemplate = `
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: workerflow
+  namespace: {{.TestNamespace}}
+spec:
+  template:
+    spec:
+      containers:
+      - name: workerflow
+        image: "prajithp/temporal-sample:1.0.0"
+        imagePullPolicy: Always
+        env:
+        - name: TEMPORAL_ADDR
+          value: {{.TemporalDeploymentName}}.{{.TestNamespace}}.svc.cluster.local:7233
+      restartPolicy: OnFailure
+  backoffLimit: 4
+`
+)
+
+func getTemplateData() (templateData, []Template) {
+	return templateData{
+			TestNamespace:          testNamespace,
+			TemporalDeploymentName: TemporalDeploymentName,
+			ScaledObjectName:       scaledObjectName,
+			DeploymentName:         deploymentName,
+		}, []Template{
+			{Name: "scaledObjectTemplate", Config: scaledObjectTemplate},
+			{Name: "deploymentTemplate", Config: deploymentTemplate},
+		}
+}
+
+func TestTemporalScaler(t *testing.T) {
+	kc := GetKubernetesClient(t)
+	data, templates := getTemplateData()
+	CreateNamespace(t, kc, testNamespace)
+
+	KubectlApplyWithTemplate(t, data, "temporalServiceTemplate", temporalServiceTemplate)
+	KubectlApplyWithTemplate(t, data, "temporalDeploymentTemplate", temporalDeploymentTemplate)
+	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, TemporalDeploymentName, testNamespace, 1, 30, 4), "temporal is not in a ready state")
+
+	KubectlApplyMultipleWithTemplate(t, data, templates)
+	testActivation(t, kc, data)
+	testScaleOut(t, kc, data)
+	testScaleIn(t, kc, data)
+	DeleteKubernetesResources(t, testNamespace, data, templates)
+}
+
+func testActivation(t *testing.T, kc *kubernetes.Clientset, data templateData) {
```

Comment:
I will work on some test cases here, but I suspect that the Temporal scaler is not going to be able to provide accurate activation (specifically 1->0). It does not have sight of activity levels that are low enough not to cause a backlog. We may need to couple the Temporal scaler with the prometheus scaler for the activation aspect for now.

Reply:
do we need to add any other test cases? @robholland

Reply:
Yeah, we'll need a build ID test, where a new deployment comes up with a new build ID and the two deployments are shown to scale separately (older one down as tasks on old version complete, newer one up as they start on that queue). I'm also not convinced that activation based on backlog is safe when scaling to 0, it won't account for tasks that are in-flight or low enough throughput to not cause a backlog. Scaling to 0 and then having to scale back up to clear the queue when things get killed and re-scheduled isn't ideal. Not sure yet the best way to test that, or whether we should just remove the activation altogether and recommend users couple the Temporal scaler with a Prometheus scaler, where Prometheus handles the activation and Temporal the scaling. Thoughts?

Reply:
Regarding activation based on backlog, I believe we should keep this option rather than removing it, as it can be partially controlled by customising scale-down behaviour in the HPA. This feature would be useful for workloads with very short execution times. @robholland

Reply:
Is there an option in omes to specify a build ID for the  test cases? @robholland

Reply:
> I'm also not convinced that activation based on backlog is safe when scaling to 0, it won't account for tasks that are in-flight or low enough throughput to not cause a backlog. Scaling to 0 and then having to scale back up to clear the queue when things get killed and re-scheduled isn't ideal. 

We actually ran into this problem while testing this new feature. Would it be possible to have a separate metric that counts not only the pending tasks in the task queue, but also any current in-flight activities? That would be a good metric to use for the activation function.

> Recommend users couple the Temporal scaler with a Prometheus scaler, where Prometheus handles the activation and Temporal the scaling.

How would that work? I thought the problem with using the Prometheus metrics for activation is that the latency metrics come from the workers, so once the workers are scaled to zero there won't be any more metrics to use for activation?

Reply:
@Prajithp BuildID wasn't added yet to Omes, I've filed a PR: https://github.com/temporalio/omes/pull/117. Makes sense to leave activation if scaling down can be delayed via HPA. Would definitely want to mention and give an example of that in the docs for the scaler.

Reply:
@jhecking For scaling from 0, backlog is fine as an activation metric. The problem is scaling to 0, where there is still load running through the queue, just low enough that there is no backlog. We are investigating making https://github.com/temporalio/api/blob/master/temporal/api/taskqueue/v1/message.proto#L138 or similar more viable for this use case, currently it does not include some activity the scaler would need to be aware of.

Reply:
Omes PR merged. So the `--build-id` argument would need passing to both deployments (with different IDs) and the server notified about the change of routing using `temporal task-queue versioning commit-build-id --task-queue my-tq --build-id $BUILD_ID` as per https://github.com/temporalio/temporal/blob/main/docs/worker-versioning.md.

Reply:
@robholland Here is one of our use-cases: We have a worker responsible for deploying new infrastructure. This worker receives tasks only occasionally. These tasks typically take a few minutes to execute, but can occasionally take much longer as well - up to about half an hour. We were planning to use Keda to spin up this worker only as needed.

Our challenge currently is that once the worker has picked a task from the task queue and started executing it, the task queue will often be empty again. So after the cool down period has passed, Keda will deactivate the worker again, even though it may still be busy executing in-flight tasks.

One obvious solution is to extend the cool down period to ensure it is longer than the longest expected activity duration. Not a great solution, but should be workable. Another solution I'm currently looking into, is whether we can intercept the termination signal sent to the worker and keep the worker running until all in-flight activities have been completed, i.e. a graceful shutdown. A third, an my preferred, option would be if there would be a separate metric that covers both pending and in-flight activities. I think the `tasks_dispatch_rate` metric _could_ be that metric, but the fact that it doesn't include sticky activities is a problem. The deployment tasks this worker handles are typically workflows that consist of a few separate activities. So only the first (usually quick) workflow task would count towards the `tasks_dispatch_rate` metric, but not the subsequent activity tasks, which will be the (somewhat) long-running tasks.

Reply:
> I think the tasks_dispatch_rate metric could be that metric.

Actually, I take that back. If it is averaged over the last 30 seconds then this metric will also quickly go to zero while this worker is executing any activity that takes longer than that. Which would again lead to the deactivation of the worker.

Reply:
@robholland I'm a bit confused about the build-ID test case. In KEDA, we can only target a single deployment, so when we update the build-ID, all pods are updated on a rolling basis. We can focus on creating test cases to verify whether the build-ID feature works as expected by testing message consumption.

Reply:
When using the new worker versioning system you don't update a deployment to a new build ID, rather you create a new deployment for the new build ID. The two deployments will both be present until the workflows on the older build ID have completed, when you tear it down. So we'd have a scaled object for each deployment/build id.

Reply:
@robholland added a test case for BuildID. Can you please verify?

Reply:
@robholland 

> I'll be investigating prometheus metrics we might be able to use to tide us over until we can provide some equivalent of `tasks_dispatch_rate` that covers in-flight (if that's going to be possible).

I know you've been out for a week, but was wondering whether you've had any time to look into this? Maybe I'm missing something, but wouldn't `tasks_dispatch_rate` also only be useful for 1->0 scaling, but not for 0->1 scaling? I.e. if all the workers have been shut down, then no new tasks can be dispatched, so the task dispatch rate cannot go back above zero.

I still think a variant of the `approximate_backlog_count` metric that includes both pending and in-flight tasks would be best suited for 1->0 and 0->1 scaling.

I'm new to KEDA myself. Is it possible to use different metrics for 1->0/0->1 and 1->n/n->1 scaling decisions? Maybe by using two separate triggers?

Reply:
Yes task dispatch rate would only be useful for de-activation (but not reliable, due to it's caveats as discussed).. `approximate_backlog_count` should be useable for activation. You can mix different triggers together, I believe you'd use https://keda.sh/docs/2.14/reference/scaledobject-spec/#scalingmodifiers for that.